src/ConstantBounds.cpp

```diff
@@ -0,0 +1,1141 @@
+    return Mul::make(std::move(recurse), std::move(expr_factor));
+}
+
+Expr handle_push_none(const Expr &expr, Direction direction) {
+    debug(3) << "push_none(" << expr << ", " << direction << ")\n";
+    // Upper bound or lower bound without a factor or denominator
+    if (const Add *op = expr.as<Add>()) {
+        return handle_push_none(op->a, direction) + handle_push_none(op->b, direction);
+    } else if (const Sub *op = expr.as<Sub>()) {
+        return handle_push_none(op->a, direction) - handle_push_none(op->b, flip(direction));
+    } else if (const Min *op = expr.as<Min>()) {
+        return min(handle_push_none(op->a, direction), handle_push_none(op->b, direction));
+    } else if (const Max *op = expr.as<Max>()) {
+        return max(handle_push_none(op->a, direction), handle_push_none(op->b, direction));
+    } else if (const Select *op = expr.as<Select>()) {
+        const Expr true_value = handle_push_none(op->true_value, direction);
+        const Expr false_value = handle_push_none(op->false_value, direction);
+        return select(op->condition, true_value, false_value);
+    } else if (const Mul *op = expr.as<Mul>()) {
+        if (const IntImm *constant = op->b.as<IntImm>()) {
+            if (constant->value == 1) {
+                // strip off a mul-by-one.
+                return handle_push_none(op->a, direction);
+            } else {
+                return handle_push_mul(op->a, direction, constant->value);
+            }
+        } else if (const IntImm *constant = op->a.as<IntImm>()) {
+            if (constant->value == 1) {
+                // strip off a mul-by-one.
+                return handle_push_none(op->b, direction);
+            } else {
+                return handle_push_mul(op->b, direction, constant->value);
+            }
+        }
+    } else if (const Div *op = expr.as<Div>()) {
+        if (const IntImm *constant = op->b.as<IntImm>()) {
+            if (constant->value > 0) {
+                return handle_push_div(op->a, direction, constant->value);
+            } else {
+                Expr recurse = handle_push_none(op->a, flip(direction));
+                return Div::make(std::move(recurse), op->b);
+            }
+        }
+    }
+    return expr;
+}
+
+// Taken from Simplify_Let.cpp
+class CountVarUses : public IRVisitor {
+    std::map<std::string, int> &var_uses;
+
+    void visit(const Variable *var) override {
+        var_uses[var->name]++;
+    }
+
+    void visit(const Load *op) override {
+        var_uses[op->name]++;
+        IRVisitor::visit(op);
+    }
+
+    void visit(const Store *op) override {
+        var_uses[op->name]++;
+        IRVisitor::visit(op);
+    }
+
+    using IRVisitor::visit;
+
+public:
+    CountVarUses(std::map<std::string, int> &_var_uses)
+        : var_uses(_var_uses) {
+    }
+};
+
+/*
+ * Visitor for removing terms that are completely unbounded from
+ * a min or a max.
+ */
+class StripUnboundedTerms : public IRMutator {
```

Comment:
It looks like this class can only correctly handle the listed IR node types, and everything it would make a mess on is explicitly listed below. It seems like this would be safer as a recursive function - when someone adds a new IR node type this class will be implicitly broken and they won't know to update it.

Reply:
I agree. I will do this, but probably won't be able to today, hopefully tomorrow.

Reply:
changed in 7a2615444. Would love your thoughts on the `Let` case, that's the one I'm a little unsure on.